### PR TITLE
Leveling Sensor Review

### PR DESCRIPTION
**Pull request by rick-vogel**

### CODE REVIEW

File: src/controls/LevelingSensor/QsgNodeUtility.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
+#pragma once
+
+#include <QSGNode>
+#include <type_traits>
+
+namespace qsg
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QsgNodeUtility.h]: That looks like a namespace reserved by the Qt company :) Maybe QskSG or so?

File: src/controls/LevelingSensor/QsgNodeUtility.cpp [hunk 2]
```diff
@@ -0,0 +1 @@
+#include "QsgNodeUtility.h"
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QsgNodeUtility.cpp]: Maybe call those files QskSG*, like we do with QskSGNode etc.? Qsg* looks like as if it was coming from the Qt Company...

File: src/controls/LevelingSensor/QskLevelingSensor.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            dst = src;
+            return true;
+        }
+        return false;
+    }
+
+    template<>
+    bool compareExchange<float>(float& dst, const float& src)
+    {
+        if (!qFuzzyCompare(dst, src))
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensor.cpp]: You could use qskFuzzyCompare here, which also works when one of the values is 0

File: src/controls/LevelingSensor/QskLevelingSensor.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    void anglesChanged(const QVector3D& degree);
+public:
+    Q_REQUIRED_RESULT const QVector3D& rotation() const noexcept;
+    Q_REQUIRED_RESULT const Tickmarks& tickmarks(Qt::Axis axis) const;
+    Q_REQUIRED_RESULT const TickmarksLabels& tickmarkLabels(Qt::Axis axis) const;
+    Q_REQUIRED_RESULT const QVector3D& angle() const noexcept;
+private:
+    /// @brief The sensors rotation per axis: x := roll, y := pitch, z := yaw
+    QVector3D m_rotation;
+    QVector3D m_angle = { 45,45,45 };
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensor.h:50]: Should they be 0 in the beginning?

File: src/controls/LevelingSensor/QskLevelingSensorNodes.h [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    }
+
+    QSGGeometry m_geometry;
+    QSGFlatColorMaterial m_material;
+    qreal m_r1 = 0.0;
+    QVector3D m_r2 = { 0.5, 0.75, 1.0 };
+    QskHashValue m_tickmarksHash{ 0 };
+};
+
+class LinearTickmarksNode final : public QSGGeometryNode
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorNodes.h:91]: Use QskTickmarksNode here? if the tick factor for the specific ticks is not settable, maybe we could have public setters for this. I guess this is something to discuss with Uwe...

File: src/controls/LevelingSensor/QskLevelingSensorNodes.h [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    }
+
+    QSGGeometry m_geometry;
+    QSGFlatColorMaterial m_material;
+    QVector2D m_scale = { 1.0f, 0.0f };
+    QVector2D m_offset = { 0.0f, 0.0f };
+    QVector3D m_tickmarkSize = { 1.0, 2.0, 4.0 };
+};
+
+class PolygonClipNode final : public QSGClipNode
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorNodes.h]: Should this be called RadialClipNode?

File: src/controls/LevelingSensor/QskLevelingSensorNodes.h [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+    QVector2D value(const QVector2D& v, const QVector2D& s, const QVector2D& o) const
+    {
+        return static_cast<const CRTP*>(this)->value(v, s, o);
+    }
+
+    void update(const QskSkinnable* const skinnable, const QskAspect::Subcontrol subControl, const QVector<QPair<double, QString>>& labels, const QVector2D& scale = { 1.0, 0.0 }, const QVector2D& offset = {})
+    {
+        if (childCount() != labels.count())
+        {
+            removeAllChildNodes();
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorNodes.h]: I think there is no need to delete all child nodes; rather add the amount of missing nodes or remove the amount of superfluous ones.

File: src/controls/LevelingSensor/QskLevelingSensorSkinlet.cpp [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+#include <QSGFlatColorMaterial>
+#include <QSGGeometryNode>
+#include <qmath.h>
+
+using Q = LevelingSensor;
+using R = LevelingSensorSkinlet::NodeRole;
+
+using namespace qsg;
+
+float LevelingSensorSkinlet::radius2(const QskSkinnable* const skinnable)
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorSkinlet.cpp]: maybe call this outerRadius?

File: src/controls/LevelingSensor/QskLevelingSensorSkinlet.cpp [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+    {
+        return subControlRect<Horizon>(sensor, contentsRect);
+    }
+
+    return Inherited::subControlRect(skinnable, contentsRect, subControl);
+}
+
+template<LevelingSensorSkinlet::NodeRole>
+QSGNode* LevelingSensorSkinlet::updateSubNode(const LevelingSensor* const sensor,
+    const quint8 nodeRole, QSGNode* const node) const = delete;
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorSkinlet.cpp]: Why the template code here? I guess standard in QSkinny is to have one common updateSubNode() and check for the node role there...

File: src/controls/LevelingSensor/QskLevelingSensorUtility.cpp [hunk 10]
```diff
@@ -0,0 +1 @@
+#include "QskLevelingSensorUtility.h"
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorUtility.cpp]: Do we need this file? Also, some other cpp file is empty

File: src/controls/LevelingSensor/QskLevelingSensorUtility.h [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    if (!qFuzzyCompare(dst, src))
+    {
+        dst = src;
+        return true;
+    }
+    return false;
+}
+
+template<>
+inline bool compare_exchange<qreal>(qreal& dst, const qreal& src)
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorUtility.h]: Qt and QSkinny uses camelCase for function names

File: src/controls/LevelingSensor/QskLevelingSensorUtility.h [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+{
+    if (!qFuzzyCompare(dst, src))
+    {
+        dst = src;
+        return true;
+    }
+    return false;
+}
+
+inline QskScaleTickmarks filtered(const QskScaleTickmarks& tickmarks, const std::function<bool(QskScaleTickmarks::TickType, qreal)>& predicate)
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorUtility.h:74]: What does this do?

File: src/controls/LevelingSensor/QskLevelingSensorUtility_test.cpp [hunk 13]
```diff
@@ -0,0 +1,190 @@
+#include <cstdlib>
+#include <type_traits>
+#include <iostream>
+#include <string>
+
+#include <QSGNode>
+
+#include "QskLevelingSensorUtility.h"
+
+#define ASSERT_TRUE(expr) if(!(expr)) { std::cerr << "ASSERT_TRUE(!" << #expr << "):" << __LINE__ << '\n'; return false; }
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorUtility_test.cpp]: I guess you could just use QVERIFY from QtTest for this, which also prints the line IIRC

File: src/controls/LevelingSensor/QskLevelingSensorUtility_test.cpp [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+        ASSERT_TRUE(dynamic_cast<QSGNodeA*>(root->childAtIndex(1)->childAtIndex(2)));
+        ASSERT_TRUE(dynamic_cast<QSGNodeG*>(root->childAtIndex(2)));
+        ASSERT_TRUE(dynamic_cast<QSGNodeC*>(root->childAtIndex(2)->childAtIndex(0)));
+        ASSERT_TRUE(dynamic_cast<QSGNodeA*>(root->childAtIndex(2)->childAtIndex(1)));
+        ASSERT_TRUE(dynamic_cast<QSGNodeB*>(root->childAtIndex(2)->childAtIndex(2)));
+    }
+    return true;
+}
+
+int main()
```
Comment (peter-ha) [inline src/controls/LevelingSensor/QskLevelingSensorUtility_test.cpp]: use the QtTest structure for this?

**Comment by rick-vogel** [comment]
PR continued here: https://github.com/uwerat/qskinny/pull/333

**Pull request closed without merging**